ExampleMod/Common/UI/ResourceOverlay/VanillaLifeOverlay.cs

```diff
@@ -0,0 +1,138 @@
+﻿using ExampleMod.Common.Players;
+using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using ReLogic.Content;
+using System.Collections.Generic;
+using Terraria;
+using Terraria.GameContent;
+using Terraria.ModLoader;
+
+namespace ExampleMod.Common.UI.ResourceOverlay
+{
+	public class VanillaLifeOverlay : ModResourceOverlay
+	{
+		// This field is used to cache vanilla assets used in the CompareAssets helper method further down in this file
+		private Dictionary<string, Asset<Texture2D>> vanillaAssetCache = new();
+
+		// These fields are used to cache the result of ModContent.Request<Texture2D>()
+		private Asset<Texture2D> heartTexture, fancyPanelTexture, barsFillingTexture, barsPanelTexture;
+
+		public override void PostDrawResource(ResourceOverlayDrawContext context) {
+			Asset<Texture2D> asset = context.texture;
+
+			string fancyFolder = "Images/UI/PlayerResourceSets/FancyClassic/";
+			string barsFolder = "Images/UI/PlayerResourceSets/HorizontalBars/";
+
+			bool drawingBarsPanels = CompareAssets(asset, barsFolder + "HP_Panel_Middle");
+
+			// Bars panel drawing has two additional elements, so the "resource index" needs to be offset
+			int fillResourceNumber = drawingBarsPanels ? context.resourceNumber - 1 : context.resourceNumber;
```

Comment:
Alright, all we need is for this value to be provided by the context, rather than calculated here.